modules/products/bazel/src/main/kotlin/com/metalbear/mirrord/products/bazel/BazelBinaryProvider.kt

```diff
@@ -0,0 +1,271 @@
+ */
+class RestoreFailed(message: String) : Exception(message) {
+    constructor(message: String, cause: Throwable) : this(message) {
+        initCause(cause)
+    }
+}
+
+/**
+ * BinaryExecutionPlan is used to group the datas needed for an execution with mirrord
+ */
+interface BinaryExecutionPlan {
+
+    /**
+     * Retrieve the original env that is present in the context of the bazel build
+     *
+     * @return the env obtained before the execution starts
+     */
+    fun getOriginalEnv(): ImmutableMap<String, String>
+
+    /**
+     * Add a list of env variables to the current env, before the execution
+     */
+    fun addToEnv(map: Map<String, String>)
+
+    /**
+     * Get the path to the binary to patch with mirrord
+     *
+     * @return the path of the target executable to patch
+     */
+    fun getBinaryToPatch(): String?
+
+    /**
+     * Check whether the state is coherent for an execution, it returns the original bazel executable used for building
+     *
+     * @return the path to the bazel executable
+     * @param executionInfo infos related to the current planned mirrord execution
+     * @throws ExecutionCheckFailed
+     */
+    @Throws(ExecutionCheckFailed::class)
+    fun checkExecution(executionInfo: MirrordExecution): String
+
+    /**
+     * Restore the environment after an execution, this method should fail if something make the old configuration
+     * restore impossible
+     *
+     * @param savedConfigData The saved environment before the execution, to be restored and the original bazel
+     * execution path
+     * @throws RestoreFailed
+     */
+    @Throws(RestoreFailed::class)
+    fun restoreConfig(savedConfigData: SavedConfigData)
+
+}
+
+/**
+ * Preserves original configuration for active Bazel runs (user environment variables and Bazel binary path)
+ */
+interface BazelBinaryProvider {
+    fun provideTargetBinaryExecPlan(executorId: String): BinaryExecutionPlan?;
+
+    fun getBinaryExecPlanClass(): KClass<out BinaryExecutionPlan>;
+
+    companion object Factory {
+
+        fun fromExecutionEnv(env: ExecutionEnvironment): BazelBinaryProvider {
+            try {
+                val clazz = Class.forName("com.google.idea.blaze.base.bazel.BuildSystem")
+                val getBuildInvokers = clazz.methods.filter { method -> method.name == "getBuildInvoker" }
+                val method253Exist =
+                    getBuildInvokers.find { method -> method.parameters.size == 1 && method.parameters[0].type.name == "Project" } != null
+                val method241Exist = getBuildInvokers.find { method ->
+                    method.parameters.size == 2 && method.parameters[0].type.name == "Project" && method.parameters[1].type.name == "BlazeContext"
+                } != null
+
+
+                if (method253Exist && method241Exist) {
+                    throw BuildExecPlanError("Unable to determine  the right bazel API version")
```

Comment:
Can't we instead of erroring, check exactly if the class has enough for us to work with, and if it does, work with that, even if it also has other stuff?